### PR TITLE
Restructure software architecture section to demonstrate refactoring and paradaigms

### PR DESCRIPTION
**Pull request by thomaskileyukaea**
This is a re-write of section 3, to address the #227. The aim is to focus on refactoring existing code, and demonstrate the principles of OOP, functional etc through examples where they are useful, rather than explicitly teaching them as concepts. The hope is that the result will be quicker and easier to teach. 

The changes rely on a new bit of code in the example project. This can be found here:

https://github.com/thomaskileyukaea/python-intermediate-inflammation/compare/thk/code-changes-for-architecture?expand=1

# ✅ ~Still to do for structural draft~

 - [x] Add headers for the new pages
 - [x] Add exercises 
 - [x] Review the removed questisons / objectives / keypoints to determine if they are still covered, or should be cut
 - [x] Review the removed content to determine if the important stuff is still covered

# ✅ ~Still to do for first draft~

- [x] Initial draft of content
    - [x] Software design
    - [x] Refactoring functions
    - [x] Refactoring architecture
    - [x] Decoupling code
    - [x] YAGNI
- [x] Update section 3 intro
- [x] Line by line read through
- [x] Solo run to iron out issues
- [x] Dummy run with another person to get a sense of timing and identify issues

# Still to do for final draft
- [ ] overlap between questions and objectives in 32 (general pass on Qs and objectives)
- [ ] Discuss removing mention of imperative code or give an example of the two side by side
- [ ] Starting point code merged into the inflammation repo (or else steps for checking out the branch)
- [ ] Careful read through

# Nice to have
- [ ] Section on DRY and inheritance - separating out the `glob` into a DataSource
- [ ] Different problem for the design - try the problem we have been doing
- [ ] Introduce an exercise for looking at the patterns (if there is time)
- [ ] Maybe don't require the imports to be inside the test functions as that is

### CODE REVIEW

**Review by thomaskileyukaea — COMMENTED** [review]
- [ ]  There is overlap between questions and objectives for lots of the episodes - would be good to understand the difference and refine this
- [ ] Don't have imports inside the test functions - is not best practise

Things to discuss:
* (how) should we mention DRY
* Should we remove the interfaces bit - seems a bit unnecessary in an untyped lang like python

**Comment by anenadic** [comment]
> * [ ]   There is overlap between questions and objectives for lots of the episodes - would be good to understand the difference and refine this

Questions are lesson objectives reworded as questions that will be answered as part of the episode but in a language that is more understandable to the audience and without the potentially unfamiliar terminology used in lesson objectives.

**Comment by thomaskileyukaea** [comment]
> > * [ ]    There is overlap between questions and objectives for lots of the episodes - would be good to understand the difference and refine this
> 
> Questions are lesson objectives reworded as questions that will be answered as part of the episode but in a language that is more understandable to the audience and without the potentially unfamiliar terminology used in lesson objectives.

That's really helpful @anenadic thank you! I won't worry too much about them sounding like they are repeating themselves, but I'll do a pass to ensure the q's are all in simple language.

File: _episodes/35-refactoring-decoupled-units.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+>> this is a good idea as otherwise this exercise is very hard, here's what we're aiming for:
+>> See commit: https://github.com/thomaskileyukaea/python-intermediate-inflammation/commit/7ccda313fda3a0b10ef5add83f5be50fe1d250fd
+>> At the end of this exercise, they would have implemented `CSVDataSource`.
+> {: .solution}
+{: .challenge}
+
+* Introduce what an interface is
+* Introduce what polymorphism is
+* Explain how we can use polymorphism to introduce abstractions
+
```
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/35-refactoring-decoupled-units.md]: Decided to instead remove interfaces as an actual thing

File: _episodes/33-refactoring-functions.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
+---
+title: "Refactoring functions to do just one thing"
```
Comment (bielsnohr) [inline _episodes/33-refactoring-functions.md:2]:
```suggestion
title: "Refactoring Functions to Do Just One Thing"
```

File: _episodes/34-refactoring-decoupled-units.md [hunk 3]
```diff
@@ -0,0 +1,443 @@
+---
+title: "Using classes to de-couple code."
```
Comment (bielsnohr) [inline _episodes/34-refactoring-decoupled-units.md:2]:
```suggestion
title: "Using Classes to De-Couple Code"
```

File: _episodes/35-refactoring-architecture.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
+---
+title: "Architecting code to separate responsibilities"
```
Comment (bielsnohr) [inline _episodes/35-refactoring-architecture.md:2]:
```suggestion
title: "Architecting Code to Separate Responsibilities"
```

**Review by bielsnohr — CHANGES_REQUESTED** [review]
Please change episode titles to respect title case used for all other episodes in course

**Review by bielsnohr — APPROVED** [review]
Thanks for the minor fixes. All good from my end. Passing on to other maintainers.

File: _episodes/30-section3-intro.md [hunk 5]
```diff
@@ -131,15 +131,10 @@ within the context of the typical software development process:
...
-  so it is worthwhile being familiar with them and knowing when
-  to switch in order to make better code.
-- How you can (and should) assess and update a software's architecture when
-  requirements change and complexity increases -
-  is the architecture still fit for purpose,
-  or are modifications and extensions becoming increasingly difficult to make?
+- How to improve existing code to be more readable, maintainable and testable.
+- Consider different strategies for writing well designed code, including
+  using **pure functions**, **classes** and **abstractions**.
+- How to create, asses and improve software design.
```
Comment (douglowe) [inline _episodes/30-section3-intro.md:137]: 'assess' rather than 'asses'?

**Comment by anenadic** [comment]
Merging to a local branch from Thomas's repo for easier review as Thomas's time on this project ended and then will reopen a new PR.

**Comment by anenadic** [comment]
> * [ ]   There is overlap between questions and objectives for lots of the episodes - would be good to understand the difference and refine this
[snip]

The questions help learners understand what to expect from an episode/lesson as they might not yet understand the learning objectives.

**Pull request merged**